packages/salesforce-adapter/src/filters/custom_object_instances_references.ts

```diff
@@ -0,0 +1,82 @@
+/*
+*                      Copyright 2020 Salto Labs Ltd.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import _ from 'lodash'
+import { transformValues, TransformFunc } from '@salto-io/adapter-utils'
+import {
+  Element, isInstanceElement, Values, ObjectType, Field, isPrimitiveType, Value, InstanceElement,
+  ReferenceExpression,
+} from '@salto-io/adapter-api'
+import { FilterCreator } from '../filter'
+import { isCustomObject, Types, apiName } from '../transformers/transformer'
+import { FIELD_ANNOTATIONS, CUSTOM_OBJECT_ID_FIELD } from '../constants'
+
+const replaceReferenceValues = (
+  values: Values,
+  type: ObjectType,
+  instances: InstanceElement[]
+): Values => {
+  const shouldReplace = (field: Field): boolean => (
+    isPrimitiveType(field.type)
+    && (
+      Types.primitiveDataTypes.Lookup.isEqual(field.type)
+      || Types.primitiveDataTypes.MasterDetail.isEqual(field.type)
+    )
+  )
+
+  const replacer = (val: Value, field: Field): Value => {
+    if (field.annotations[FIELD_ANNOTATIONS.REFERENCE_TO] === undefined
+        || !_.isArray(field.annotations[FIELD_ANNOTATIONS.REFERENCE_TO])) {
+      return val
+    }
+    const refToInstance = instances.find(instance =>
+      field.annotations[FIELD_ANNOTATIONS.REFERENCE_TO].includes(apiName(instance.type, true))
+      && instance.value[CUSTOM_OBJECT_ID_FIELD] === val)
+    return (refToInstance === undefined) ? val : new ReferenceExpression(refToInstance.elemID)
+  }
+
+  const transformFunc: TransformFunc = ({ value, field }) => (
+    !_.isUndefined(field) && shouldReplace(field) ? replacer(value, field) : value
+  )
```

Comment:
probably worth indexing instances by type and ID instead of searching all instances for each reference.

also, the implementation can be so much shorter
```suggestion
  const replacer: TransformFunc = ({ value, field }) => {
    if (!shouldReplace(field)) { // we don't really need this part, but it is nice to have I guess... up to you
      return value
    }
    const refTo = makeArray(field?.annotations?.[FIELD_ANNOTATIONS.REFERENCE_TO])
    const refTarget = refTo.map(typeName => instancesByType[typeName].?[value]).filter(values.isDefined).pop()
    return refTarget === undefined ? val : new ReferenceExpression(refTarget.elemID)
  }
```